Don't let users watch Web Sessions with secrets

Joerger opened the pull request:
Users are currently permitted to set up watchers on their own web sessions. This is used by clients to check when a web session is propagated through the cache and ready to support connections. In these cases the user does not actually need the web session secrets. At most they need the web session ID.

This PR adds logic for `kind.LoadSecrets` when watching web sessions, and prevents users from watching their own secrets. This is particularly important for Hardware Key support as the expectation is that web session secrets are only [known by the Proxy and Auth services](https://github.com/gravitational/teleport/blob/master/rfd/0080-hardware-key-support.md#web-session-access).

Note: This is not really a security fix as an attacker would still need access to the user's hardware key and MFA device to carry out an attack. This just hardens Hardware Key support for the cluster in general.

Changelog: Harden web session security for Hardware Key support

Joerger committed:
Don't watch web sessions with secrets by default; Deny users from watching web session secrets.

Joerger commented:
@tcsc @fspmarshall friendly reminder to review

fspmarshall reviewed (approved):
This needs backwards-compatibility logic to ensure that auth servers running this logic don't break older proxies.  If an auth receives read/watch requests from older proxies it should modify the request inflight to set `loadSecrets` to true.  LGTM otherwise.

Joerger committed:
Add backwards compatibility logic for old proxies

Fallback to web session permissions for users watching with secrets

Fix uncaught merge conflict

Joerger committed:
CopyWithoutSecrets without proto.Clone to avoid gogoproto issue.

Joerger committed:
Revert WebSession.CopyWithoutSecrets rename and instead add special case in fanoutv2.go since WebSessions can not implement ResourceWithSecrets without affecting upstream consumers negatively.

Joerger commented on api/types/session.go:
```diff
@@ -86,8 +85,8 @@ type WebSession interface {
-	// WithoutSecrets returns copy of the web session but without private keys
-	WithoutSecrets() WebSession
+	// CopyWithoutSecrets returns copy of the web session but without private keys
+	CopyWithoutSecrets() WebSession
```
addressed in 7e0a738a77108407745943e482156aa8f6dc2cbf

Joerger committed:
Address slice iteration comment.

Joerger commented on api/types/session.go:
```diff
@@ -194,18 +193,11 @@ func (ws *WebSessionV2) GetIdleTimeout() time.Duration {
-	cp := proto.Clone(ws).(*WebSessionV2)
+	cp := *ws
 	cp.Spec.Priv = nil
 	cp.Spec.SAMLSession = nil
 	cp.Spec.DeviceWebToken = nil
-	return cp
+	return &cp
```
No I think this was an uncaught merge conflict, thanks for catching it.

Joerger replied:
Ah no I remember now, it was causing race conditions in tests when the times are changed to UTC, which is needed to make proto.Clone work with gogoproto.

Joerger committed:
Merge branch 'master' into joerger/user-cant-watch-web-session-secrets

Joerger committed:
Merge branch 'master' into joerger/user-cant-watch-web-session-secrets

Joerger committed:
Merge branch 'master' into joerger/user-cant-watch-web-session-secrets
